demonstrations/tutorial_fdhs.py

```diff
@@ -0,0 +1,455 @@
+
+
+##############################################################################
+# We have successfully decomposed the :math:`60`-dimensional Lie algebra 
+# into a :math:`24`-dimensional vertical subspace and a :math:`36`-dimensional subspace.
+#
+# Note that not every bipartition of a Lie algebra constitutes a Cartan decomposition.
+# For that, the subspaces need to fulfil the following three commutation relations
+#
+# .. math::
+#     \begin{align}
+#     [\mathfrak{k}, \mathfrak{k}] \subseteq \mathfrak{k} & \text{ (subalgebra)}\\
+#     [\mathfrak{k}, \mathfrak{m}] \subseteq \mathfrak{m} & \text{ (reductive property)}\\
+#     [\mathfrak{m}, \mathfrak{m}] \subseteq \mathfrak{k} & \text{ (symmetric property)}
+#     \end{align}
+#
+# In particular, :math:`\mathfrak{k}` is closed under commutation and is therefore a subalgebra, whereas :math:`\mathfrak{m}` is not.
+# This also has the consequence that the associated Lie group :math:`\mathcal{K} := e^{i \mathfrak{k}}` is a subgroup
+# of the associated Lie group :math:`\mathcal{G} := e^{i \mathfrak{g}}`.
+#
+# Cartan subalgebra
+# -----------------
+# 
+# With this we have identified the first subgroup (:math:`\mathcal{K}`) of the KAK decomposition. The other subgroup
+# is induced by the so-called (horizontal) Cartan subalgebra :math:`\mathfrak{h}`. This is a maximal Abelian subalgebra of :math:`\mathfrak{m}` and is not unique.
+# For the case of Pauli words we can simply pick any element in :math:`\mathfrak{m}` and collect all other operators in :math:`\mathfrak{m}`
+# that commute with it.
+#
+# We then obtain a further split of the vector space :math:`\mathfrak{m} = \tilde{\mathfrak{m}} \oplus \mathfrak{h}`,
+# where :math:`\tilde{\mathfrak{m}}` is just the remainder of :math:`\mathfrak{m}`.
+
+def _commutes_with_all(candidate, ops):
+    """Check if ``candidate commutes with all ``ops``"""
+    for op in ops:
+        com = candidate.commutator(op)
+        com.simplify()
+        
+        if not len(com) == 0:
+            return False
+    return True
+
+def cartan_subalgebra(m, which=0):
+    """Compute the Cartan subalgebra from the odd parity space :math:`\mathfrak{m}` of the Cartan decomposition
+
+    This implementation is specific for cases of bases of m with pure Pauli words as detailed in Appendix C in `2104.00728 <https://arxiv.org/abs/2104.00728>`__.
+    
+    Args:
+        m (List[PauliSentence]): the odd parity subspace :math:`\Theta(\mathfrak{m}) = \mathfrak{m}
+        which (int): Choice for initial element of m from which to construct the maximal Abelian subalgebra
+    
+    Returns:
+        mtilde (List): remaining elements of :math:`\mathfrak{m}` s.t. :math:`\mathfrak{m} = \tilde{\mathfrak{m}} \oplus \mathfrak{h}`.
+        a (List): Cartan subalgebra
+
+    """
+
+    h = [m[which]] # first candidate
+    mtilde = m.copy()
+
+    for m_i in m:
+        if _commutes_with_all(m_i, h):
+            if m_i not in h:
+                h.append(m_i)
```

Comment:
[super optional reduction]
```suggestion
    m.remove(h[0])
    for m_i in m:
        if _commutes_with_all(m_i, h):
            h.append(m_i)
```